src/lib/precondition.ts

```diff
@@ -0,0 +1,375 @@
+import {
+  Circuit,
+  AsFieldElements,
+  Bool,
+  Field,
+  jsLayout,
+  Types,
+} from '../snarky';
+import { circuitValueEquals } from './circuit_value';
+import { PublicKey } from './signature';
+import * as Mina from './mina';
+import { Party, Preconditions } from './party';
+import * as GlobalContext from './global-context';
+import { UInt32, UInt64 } from './int';
+
+export {
+  preconditions,
+  Account,
+  Network,
+  assertPreconditionInvariants,
+  AccountValue,
+  NetworkValue,
+};
+
+function preconditions(party: Party, isSelf: boolean) {
+  initializePreconditions(party, isSelf);
+  return { account: Account(party), network: Network(party) };
+}
+
+// note: please keep the two precondition implementations separate
+// so we can add customized fields easily
+
+function Network(party: Party): Network {
+  // TODO there should be a less error-prone way of typing js layout
+  // e.g. separate keys list and value object, so that we can access by key
+  let layout = (jsLayout as any).Party.layout[0].value.layout[9].value.layout[0]
```

Comment:
the `jsLayout` provides the necessary info to produce `precondition` interfaces of the right structure at runtime